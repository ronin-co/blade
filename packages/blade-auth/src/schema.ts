import { boolean, date, link, model, string } from 'blade/schema';

export const Account = model({
  slug: 'account',
  fields: {
    email: string({ unique: true }),
    emailVerified: boolean(),
    emailVerificationToken: string(),
    emailVerificationSentAt: date(),
    password: string(),
  },
});

export const Session = model({
  slug: 'session',
  fields: {
    account: link({
      target: 'account',
      actions: {
        onDelete: 'CASCADE',
      },
    }),
    browser: string(),
    browserVersion: string(),
    os: string(),
    osVersion: string(),
    deviceType: string(),
  },
});
