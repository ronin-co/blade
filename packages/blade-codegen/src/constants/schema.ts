import { SyntaxKind, factory } from 'typescript';

import { identifiers } from '@/src/constants/identifiers';

import type { TypeNode } from 'typescript';

import type { ModelField } from '@/src/types/model';
import type { DML_QUERY_TYPES } from 'blade-compiler';

/**
 * A list of all model field types & their TypeScript type mapping.
 */
export const MODEL_TYPE_TO_SYNTAX_KIND_KEYWORD = {
  blob: factory.createTypeReferenceNode(identifiers.compiler.storedObject),
  boolean: factory.createKeywordTypeNode(SyntaxKind.BooleanKeyword),
  date: factory.createTypeReferenceNode(identifiers.primitive.date),
  json: factory.createUnionTypeNode([
    factory.createTypeReferenceNode(identifiers.utils.jsonObject),
    factory.createTypeReferenceNode(identifiers.utils.jsonArray),
  ]),
  link: factory.createKeywordTypeNode(SyntaxKind.UnknownKeyword),
  number: factory.createKeywordTypeNode(SyntaxKind.NumberKeyword),
  string: factory.createKeywordTypeNode(SyntaxKind.StringKeyword),
} satisfies Record<ModelField['type'], TypeNode>;

/**
 * A simple object mapping all DML query types to their human readable string.
 */
export const READABLE_DML_QUERY_TYPES = {
  add: 'Add',
  count: 'Count',
  get: 'Get',
  remove: 'Remove',
  set: 'Set',
  use: 'Get',
} satisfies Record<(typeof DML_QUERY_TYPES)[number] | 'use', string>;

/**
 * A list of all default field slugs for RONIN models.
 *
 * This is designed to match the properties as part of `ResultRecord`.
 */
export const DEFAULT_FIELD_SLUGS = [
  'id',
  'ronin.createdAt',
  'ronin.createdBy',
  'ronin.updatedAt',
  'ronin.updatedBy',
] satisfies Array<string>;
