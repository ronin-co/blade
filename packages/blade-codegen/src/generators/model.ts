import { SyntaxKind, factory } from 'typescript';

import { identifiers, typeArgumentIdentifiers } from '@/src/constants/identifiers';
import { DEFAULT_FIELD_SLUGS } from '@/src/constants/schema';
import { triggerOptionsInterface } from '@/src/declarations';
import { convertToPascalCase } from '@/src/utils/slug';
import { mapRoninFieldToTypeNode, remapNestedFields } from '@/src/utils/types';

import type { TypeAliasDeclaration, TypeParameterDeclaration } from 'typescript';

import type { ModelField } from '@/src/types/model';
import type { PopulatedModel } from 'blade-compiler';

/**
 * Generate all required type definitions for a provided schema model.
 *
 * The plural model type will be mapped to an array of the singular model type and
 * extended with the plural model properties.
 *
 * @param models - All models of the addressed space.
 *
 * @returns An array of type nodes to be added to the `index.d.ts` file.
 */
export const generateModelTypes = (
  models: Array<PopulatedModel>,
): Array<TypeAliasDeclaration> => {
  const nodes = new Array<TypeAliasDeclaration>(triggerOptionsInterface);

  for (const model of models) {
    const fields: Array<ModelField> = Object.entries(model.fields)
      .map(([slug, field]) => ({ ...field, slug }) as ModelField)
      .filter((field) => !DEFAULT_FIELD_SLUGS.includes(field.slug));

    const modelIdentifier = {
      singular: factory.createIdentifier(convertToPascalCase(model.slug)),
      plural: factory.createIdentifier(convertToPascalCase(model.pluralSlug)),
    };

    const hasLinkFields = fields.some(
      (field) =>
        field.type === 'link' && models.some((model) => model.slug === field.target),
    );

    const modelInterfaceTypeParameters = new Array<TypeParameterDeclaration>();
    const linkFieldKeys = fields
      .filter((field) => field.type === 'link')
      .map((field) =>
        factory.createLiteralTypeNode(factory.createStringLiteral(field.slug)),
      );

    /**
     * ```ts
     * <U extends Array<...> | 'all' = []>
     * ```
     */
    const usingGenericDec = factory.createTypeParameterDeclaration(
      undefined,
      typeArgumentIdentifiers.using,
      factory.createUnionTypeNode([
        factory.createTypeReferenceNode(identifiers.primitive.array, [
          factory.createUnionTypeNode(linkFieldKeys),
        ]),
        factory.createLiteralTypeNode(factory.createStringLiteral('all')),
      ]),
      factory.createTupleTypeNode([]),
    );

    if (hasLinkFields) modelInterfaceTypeParameters.push(usingGenericDec);

    /**
     * ```ts
     * SchemaSlugSchema<U>
     * ```
     */
    const modelSchemaName = factory.createTypeReferenceNode(
      modelIdentifier.singular,
      hasLinkFields
        ? [factory.createTypeReferenceNode(typeArgumentIdentifiers.using)]
        : [],
    );

    /**
     * ```ts
     * export type SchemaSlug<U extends Array<...> | 'all' = []> = ResultRecord & {
     *  // ...
     * };
     * ```
     */
    const singularModelTypeDec = factory.createTypeAliasDeclaration(
      [factory.createModifier(SyntaxKind.ExportKeyword)],
      modelIdentifier.singular,
      modelInterfaceTypeParameters,
      factory.createIntersectionTypeNode([
        factory.createTypeReferenceNode(
          factory.createQualifiedName(
            identifiers.namespace.utils.name,
            identifiers.namespace.utils.resultRecord,
          ),
        ),
        factory.createTypeLiteralNode(
          remapNestedFields(fields)
            .sort(([a], [b]) => a.localeCompare(b))
            .map(([fieldSlug, field]) => {
              if (Array.isArray(field)) {
                const sortedFields = field.sort((a, b) => a.slug.localeCompare(b.slug));
                return factory.createPropertySignature(
                  undefined,
                  fieldSlug,
                  undefined,
                  factory.createTypeLiteralNode(
                    sortedFields.map((nestedField) =>
                      factory.createPropertySignature(
                        undefined,
                        nestedField.slug,
                        undefined,
                        factory.createUnionTypeNode(
                          mapRoninFieldToTypeNode(nestedField, models),
                        ),
                      ),
                    ),
                  ),
                );
              }

              return factory.createPropertySignature(
                undefined,
                fieldSlug,
                undefined,
                factory.createUnionTypeNode(mapRoninFieldToTypeNode(field, models)),
              );
            }),
        ),
      ]),
    );

    /**
     * ```ts
     * Array<SchemaSlug>;
     * ```
     */
    const pluralModelArrayTypeDec = factory.createTypeReferenceNode(
      identifiers.primitive.array,
      [modelSchemaName],
    );

    /**
     * ```ts
     * {
     *  moreBefore?: string;
     *  moreAfter?: string;
     * };
     * ```
     */
    const pluralModelPaginationPropsTypeDec = factory.createTypeLiteralNode([
      factory.createPropertySignature(
        undefined,
        factory.createIdentifier('moreBefore'),
        factory.createToken(SyntaxKind.QuestionToken),
        factory.createKeywordTypeNode(SyntaxKind.StringKeyword),
      ),
      factory.createPropertySignature(
        undefined,
        factory.createIdentifier('moreAfter'),
        factory.createToken(SyntaxKind.QuestionToken),
        factory.createKeywordTypeNode(SyntaxKind.StringKeyword),
      ),
    ]);

    /**
     * ```ts
     * export type SchemaPluralSlug<U extends Array<...> | 'all' = []> = Array<SchemaSlug> & {
     *  moreBefore?: string;
     *  moreAfter?: string;
     * };
     * ```
     */
    const pluralModelTypeDec = factory.createTypeAliasDeclaration(
      [factory.createModifier(SyntaxKind.ExportKeyword)],
      modelIdentifier.plural,
      modelInterfaceTypeParameters,
      factory.createIntersectionTypeNode([
        pluralModelArrayTypeDec,
        pluralModelPaginationPropsTypeDec,
      ]),
    );

    nodes.push(singularModelTypeDec, pluralModelTypeDec);
  }

  return nodes;
};
