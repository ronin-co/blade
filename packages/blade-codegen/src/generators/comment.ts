import { READABLE_DML_QUERY_TYPES } from '@/src/constants/schema';

import type { QueryType } from '@/src/types/query';
import type { PopulatedModel } from 'blade-compiler';

interface GenerateQueryTypeCommentResult {
  singular: string;
  plural: string;
}

/**
 * Generate a text comment for a provided model & a given query type.
 *
 * @param modelName - The name of the model to generate a comment for.
 * @param queryType - The query type to generate a comment for.
 *
 * @returns An object containing both the singular and plural comment strings.
 */
export const generateQueryTypeComment = (
  model: PopulatedModel,
  queryType: QueryType,
): GenerateQueryTypeCommentResult => ({
  singular: `* ${READABLE_DML_QUERY_TYPES[queryType]} a single ${model.name ?? model.slug} record `,
  plural: `* ${READABLE_DML_QUERY_TYPES[queryType]} multiple ${model.name ?? model.slug} records `,
});
