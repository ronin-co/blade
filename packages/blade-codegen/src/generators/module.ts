import { NodeFlags, SyntaxKind, addSyntheticLeadingComment, factory } from 'typescript';

import { identifiers, typeArgumentIdentifiers } from '@/src/constants/identifiers';
import { DEFAULT_FIELD_SLUGS } from '@/src/constants/schema';
import { INFERRED_COMBINED_INSTRUCTION_PROPERTIES } from '@/src/constants/syntax';
import { generateQueryTypeComment } from '@/src/generators/comment';
import { convertToPascalCase } from '@/src/utils/slug';
import { mapRoninFieldToTypeNode } from '@/src/utils/types';

import type { Model, ModelField } from '@/src/types/model';
import type { DML_QUERY_TYPES } from 'blade-compiler';
import type {
  PropertySignature,
  TypeElement,
  TypeNode,
  VariableStatement,
} from 'typescript';

/**
 * Generate a `declare const` statement for a provided query type using a list
 * of provided models.
 *
 * @example
 * ```ts
 * declare const use: {
 *    user: ReducedFunction & {
 *      <T = User | null>(options?: Partial<CombinedInstructions>): T;
 *      after: <T = User | null>(value: CombinedInstructions["after"]) => T;
 *      before: <T = User | null>(value: CombinedInstructions["before"]) => T;
 *      including: <T = User | null>(value: CombinedInstructions["including"]) => T;
 *      limitedTo: <T = User | null>(value: CombinedInstructions["limitedTo"]) => T;
 *      orderedBy: <T = User | null>(value: CombinedInstructions["orderedBy"]) => T;
 *      selecting: <T = User | null>(value: CombinedInstructions["selecting"]) => T;
 *      using: <T = User | null>(value: CombinedInstructions["using"]) => T;
 *      with: ReducedFunction & {
 *         <T = User | null>(options: CombinedInstructions["with"]): T;
 *         id: <T = User | null>(value: ResultRecord["id"]) => T;
 *         ronin: {
 *          createdAt: <T = User | null>(value: ResultRecord["ronin.createdAt"]) => T;
 *          createdBy: <T = User | null>(value: ResultRecord["ronin.createdBy"]) => T;
 *          updatedAt: <T = User | null>(value: ResultRecord["ronin.updatedAt"]) => T;
 *          updatedBy: <T = User | null>(value: ResultRecord["ronin.updatedBy"]) => T;
 *         };
 *         name: <T = User | null>(value: string) => T;
 *         email: <T = User | null>(value: string) => T;
 *         // [...]
 *      };
 *    };
 *    users: ReducedFunction & {
 *      <T = Users>(options?: Partial<CombinedInstructions>): T;
 *      after: <T = Users>(value: CombinedInstructions["after"]) => T;
 *      before: <T = Users>(value: CombinedInstructions["before"]) => T;
 *      including: <T = Users>(value: CombinedInstructions["including"]) => T;
 *      limitedTo: <T = Users>(value: CombinedInstructions["limitedTo"]) => T;
 *      orderedBy: <T = Users>(value: CombinedInstructions["orderedBy"]) => T;
 *      selecting: <T = Users>(value: CombinedInstructions["selecting"]) => T;
 *      using: <T = Users>(value: CombinedInstructions["using"]) => T;
 *      with: ReducedFunction & {
 *         <T = Users>(options: CombinedInstructions["with"]): T;
 *         id: <T = Users>(value: ResultRecord["id"]) => T;
 *         ronin: {
 *           createdAt: <T = Users>(value: ResultRecord["ronin.createdAt"]) => T;
 *           createdBy: <T = Users>(value: ResultRecord["ronin.createdBy"]) => T;
 *           updatedAt: <T = Users>(value: ResultRecord["ronin.updatedAt"]) => T;
 *           updatedBy: <T = Users>(value: ResultRecord["ronin.updatedBy"]) => T;
 *         };
 *         name: <T = Users>(value: string) => T;
 *         email: <T = Users>(value: string) => T;
 *         // [...]
 *      };
 *    };
 * };
 * ```
 *
 * @param models - An array of RONIN models to generate query declarations for.
 * @param queryType - The type of query to generate (e.g. 'use').
 *
 * @returns A variable statement for the generated query declarations.
 */
export const generateQueryDeclarationStatements = (
  models: Array<Model>,
  queryType: (typeof DML_QUERY_TYPES)[number] | 'use',
): VariableStatement =>
  factory.createVariableStatement(
    undefined,
    factory.createVariableDeclarationList(
      [
        factory.createVariableDeclaration(
          queryType,
          undefined,
          factory.createTypeLiteralNode(
            models.flatMap((model) => {
              const comment = generateQueryTypeComment(model, queryType);

              const singularModelProperty = generateSchemaProperty(
                model.slug,
                model.fields,
                models,
                true,
              );
              const pluralModelProperty = generateSchemaProperty(
                model.pluralSlug,
                model.fields,
                models,
                false,
              );

              return [
                addSyntheticLeadingComment(
                  singularModelProperty,
                  SyntaxKind.MultiLineCommentTrivia,
                  comment.singular,
                  true,
                ),
                addSyntheticLeadingComment(
                  pluralModelProperty,
                  SyntaxKind.MultiLineCommentTrivia,
                  comment.plural,
                  true,
                ),
              ];
            }),
          ),
        ),
      ],
      NodeFlags.Const,
    ),
  );

/**
 * @todo(@nurodev): Add documentation
 */
const generateSchemaProperty = (
  modelSlug: string,
  modelFields: Model['fields'],
  models: Array<Model>,
  isNullable: boolean,
) => {
  const modelIdentifierTypes = new Array<TypeNode>(
    factory.createTypeReferenceNode(convertToPascalCase(modelSlug)),
  );
  if (isNullable)
    modelIdentifierTypes.push(factory.createLiteralTypeNode(factory.createNull()));
  const modelIdentifier = factory.createUnionTypeNode(modelIdentifierTypes);

  /**
   * ```ts
   * <T = Account | null>(options?: Partial<CombinedInstructions>): T;
   * ```
   */
  const rootInstructionHandler = generateRootInstructionHandler(
    modelIdentifier,
    factory.createTypeReferenceNode(typeArgumentIdentifiers.default, undefined),
  );

  /**
   * ```ts
   * after: <T = User | null>(value: CombinedInstructions['after']) => T ;
   * before: <T = User | null>(value: CombinedInstructions['before'],) => T ;
   * including: <T = User | null>(value: CombinedInstructions['including']) => T ;
   * limitedTo: <T = User | null>(value: CombinedInstructions['limitedTo']) => T ;
   * orderedBy: <T = User | null>(value: CombinedInstructions['orderedBy']) => T ;
   * selecting: <T = User | null>(value: CombinedInstructions['selecting']) => T ;
   * using: <T = User | null>(value: CombinedInstructions['using']) => T ;
   * ```
   */
  const mappedCombinedInstructionProperties =
    INFERRED_COMBINED_INSTRUCTION_PROPERTIES.map((propertyName) =>
      factory.createPropertySignature(
        undefined,
        propertyName,
        undefined,
        factory.createFunctionTypeNode(
          [
            factory.createTypeParameterDeclaration(
              undefined,
              typeArgumentIdentifiers.default,
              undefined,
              modelIdentifier,
            ),
          ],
          [
            factory.createParameterDeclaration(
              undefined,
              undefined,
              'value',
              undefined,
              factory.createIndexedAccessTypeNode(
                factory.createTypeReferenceNode(
                  identifiers.compiler.combinedInstructions,
                  undefined,
                ),
                factory.createLiteralTypeNode(factory.createStringLiteral(propertyName)),
              ),
              undefined,
            ),
          ],
          factory.createTypeReferenceNode(typeArgumentIdentifiers.default, undefined),
        ),
      ),
    );

  return factory.createPropertySignature(
    undefined,
    modelSlug,
    undefined,
    factory.createIntersectionTypeNode([
      factory.createExpressionWithTypeArguments(
        identifiers.utils.reducedFunction,
        undefined,
      ),
      factory.createTypeLiteralNode([
        rootInstructionHandler,
        ...mappedCombinedInstructionProperties,
        generateWithPropertySignature(
          modelIdentifier,
          modelFields,
          models,
          factory.createTypeReferenceNode(typeArgumentIdentifiers.default, undefined),
        ),
      ]),
    ]),
  );
};

/**
 * @todo(@nurodev): Add documentation
 */
const generateRootInstructionHandler = (
  modelIdentifier: TypeNode,
  returnTypeNode: TypeNode,
) =>
  factory.createCallSignature(
    [
      factory.createTypeParameterDeclaration(
        undefined,
        typeArgumentIdentifiers.default,
        undefined,
        modelIdentifier,
      ),
    ],
    [
      factory.createParameterDeclaration(
        undefined,
        undefined,
        'options',
        factory.createToken(SyntaxKind.QuestionToken),
        factory.createTypeReferenceNode(identifiers.primitive.partial, [
          factory.createTypeReferenceNode(
            identifiers.compiler.combinedInstructions,
            undefined,
          ),
        ]),
        undefined,
      ),
    ],
    returnTypeNode,
  );

/**
 * @todo(@nurodev): Add documentation
 */
const generateWithPropertySignature = (
  modelIdentifier: TypeNode,
  modelFields: Model['fields'],
  models: Array<Model>,
  returnTypeNode: TypeNode,
): PropertySignature => {
  /**
   * ```ts
   * <T = User | null>(options: CombinedInstructions["with"]): T;
   * ```
   */
  const rootInstructionHandler = factory.createCallSignature(
    [
      factory.createTypeParameterDeclaration(
        undefined,
        typeArgumentIdentifiers.default,
        undefined,
        modelIdentifier,
      ),
    ],
    [
      factory.createParameterDeclaration(
        undefined,
        undefined,
        'options',
        undefined,
        factory.createIndexedAccessTypeNode(
          factory.createTypeReferenceNode(
            identifiers.compiler.combinedInstructions,
            undefined,
          ),
          factory.createLiteralTypeNode(factory.createStringLiteral('with')),
        ),
        undefined,
      ),
    ],
    returnTypeNode,
  );

  const members = new Array<TypeElement>(
    rootInstructionHandler,
    ...generateDefaultFieldWithQueries(modelIdentifier, returnTypeNode),
  );

  for (const [slug, field] of Object.entries(modelFields)) {
    if (DEFAULT_FIELD_SLUGS.some((field) => field.includes(slug))) continue;

    members.push(
      factory.createPropertySignature(
        undefined,
        slug,
        undefined,
        factory.createFunctionTypeNode(
          [
            factory.createTypeParameterDeclaration(
              undefined,
              typeArgumentIdentifiers.default,
              undefined,
              modelIdentifier,
            ),
          ],
          [
            factory.createParameterDeclaration(
              undefined,
              undefined,
              'value',
              undefined,
              factory.createUnionTypeNode(
                mapRoninFieldToTypeNode(field as ModelField, models),
              ),
              undefined,
            ),
          ],
          returnTypeNode,
        ),
      ),
    );
  }

  return factory.createPropertySignature(
    undefined,
    'with',
    undefined,
    factory.createIntersectionTypeNode([
      factory.createExpressionWithTypeArguments(
        identifiers.utils.reducedFunction,
        undefined,
      ),
      factory.createTypeLiteralNode(members),
    ]),
  );
};

/**
 * @todo(@nurodev): Add documentation
 */
const generateDefaultFieldWithQueries = (
  modelIdentifier: TypeNode,
  returnTypeNode: TypeNode,
): Array<TypeElement> => {
  const members = new Array<TypeElement>();

  const topLevelFields = DEFAULT_FIELD_SLUGS.filter((slug) => !slug.startsWith('ronin.'));
  for (const slug of topLevelFields) {
    members.push(
      factory.createPropertySignature(
        undefined,
        slug,
        undefined,
        factory.createFunctionTypeNode(
          [
            factory.createTypeParameterDeclaration(
              undefined,
              typeArgumentIdentifiers.default,
              undefined,
              modelIdentifier,
            ),
          ],
          [
            factory.createParameterDeclaration(
              undefined,
              undefined,
              'value',
              undefined,
              factory.createIndexedAccessTypeNode(
                factory.createTypeReferenceNode(
                  identifiers.syntax.resultRecord,
                  undefined,
                ),
                factory.createLiteralTypeNode(factory.createStringLiteral(slug)),
              ),
              undefined,
            ),
          ],
          returnTypeNode,
        ),
      ),
    );
  }

  const metaFields = DEFAULT_FIELD_SLUGS.filter((slug) => slug.startsWith('ronin.'));
  const metaFieldsPropertySignatures = metaFields.map((slug) => {
    const normalizedSlug = slug.replaceAll('ronin.', '');

    return factory.createPropertySignature(
      undefined,
      normalizedSlug,
      undefined,
      factory.createFunctionTypeNode(
        [
          factory.createTypeParameterDeclaration(
            undefined,
            typeArgumentIdentifiers.default,
            undefined,
            modelIdentifier,
          ),
        ],
        [
          factory.createParameterDeclaration(
            undefined,
            undefined,
            'value',
            undefined,
            factory.createIndexedAccessTypeNode(
              factory.createIndexedAccessTypeNode(
                factory.createTypeReferenceNode(identifiers.syntax.resultRecord),
                factory.createLiteralTypeNode(factory.createStringLiteral('ronin')),
              ),
              factory.createLiteralTypeNode(factory.createStringLiteral(normalizedSlug)),
            ),
            undefined,
          ),
        ],
        returnTypeNode,
      ),
    );
  });
  members.push(
    factory.createPropertySignature(
      undefined,
      'ronin',
      undefined,
      factory.createIntersectionTypeNode([
        factory.createExpressionWithTypeArguments(
          identifiers.utils.reducedFunction,
          undefined,
        ),
        factory.createTypeLiteralNode(metaFieldsPropertySignatures),
      ]),
    ),
  );

  return members;
};
