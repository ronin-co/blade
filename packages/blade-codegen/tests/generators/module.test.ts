import { describe, expect, test } from 'bun:test';
import { model, string } from 'blade-syntax/schema';
import { NodeFlags, SyntaxKind, factory } from 'typescript';

import { generateTypedQueryMembers } from '@/src/generators/module';
import { printNodes } from '@/src/utils/print';

import type { Model } from '@/src/types/model';

describe('module', () => {
  describe('query declaration constant', () => {
    test('a basic model', () => {
      const AccountModel = model({
        slug: 'account',
        pluralSlug: 'accounts',
        fields: {
          name: string(),
          email: string({ required: true }),
        },
      });

      const models = [AccountModel] as unknown as Array<Model>;
      const members = generateTypedQueryMembers(models, 'get');
      const statement = factory.createVariableStatement(
        [factory.createModifier(SyntaxKind.DeclareKeyword)],
        factory.createVariableDeclarationList(
          [
            factory.createVariableDeclaration(
              'get',
              undefined,
              factory.createTypeLiteralNode(members),
            ),
          ],
          NodeFlags.Const,
        ),
      );
      const statementsStr = printNodes([statement]);
      expect(statementsStr).toMatchSnapshot();
    });

    test('with no modules', () => {
      const members = generateTypedQueryMembers([], 'get');
      const statement = factory.createVariableStatement(
        [factory.createModifier(SyntaxKind.DeclareKeyword)],
        factory.createVariableDeclarationList(
          [
            factory.createVariableDeclaration(
              'get',
              undefined,
              factory.createTypeLiteralNode(members),
            ),
          ],
          NodeFlags.Const,
        ),
      );
      const statementsStr = printNodes([statement]);
      expect(statementsStr).toMatchSnapshot();
    });

    test('with multiple models', () => {
      const AccountModel = model({
        slug: 'account',
        pluralSlug: 'accounts',
        fields: {
          name: string(),
          email: string({ required: true }),
        },
      });
      const PostModel = model({
        slug: 'post',
        pluralSlug: 'posts',
        fields: {
          title: string({ required: true }),
          description: string(),
        },
      });

      const models = [AccountModel, PostModel] as unknown as Array<Model>;
      const members = generateTypedQueryMembers(models, 'get');
      const statement = factory.createVariableStatement(
        [factory.createModifier(SyntaxKind.DeclareKeyword)],
        factory.createVariableDeclarationList(
          [
            factory.createVariableDeclaration(
              'get',
              undefined,
              factory.createTypeLiteralNode(members),
            ),
          ],
          NodeFlags.Const,
        ),
      );
      const statementsStr = printNodes([statement]);
      expect(statementsStr).toMatchSnapshot();
    });
  });
});
