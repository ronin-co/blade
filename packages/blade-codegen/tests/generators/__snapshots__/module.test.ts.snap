// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`module query declaration constant a basic model 1`] = `
"const get: {
    /* Get a single account record */
    account: ReducedFunction & {
        <T = Account | null>(options?: Partial<CombinedInstructions>): T;
        after: <T = Account | null>(value: CombinedInstructions["after"]) => T;
        before: <T = Account | null>(value: CombinedInstructions["before"]) => T;
        including: <T = Account | null>(value: CombinedInstructions["including"]) => T;
        limitedTo: <T = Account | null>(value: CombinedInstructions["limitedTo"]) => T;
        orderedBy: <T = Account | null>(value: CombinedInstructions["orderedBy"]) => T;
        selecting: <T = Account | null>(value: CombinedInstructions["selecting"]) => T;
        using: <T = Account | null>(value: CombinedInstructions["using"]) => T;
        with: ReducedFunction & {
            <T = Account | null>(options: CombinedInstructions["with"]): T;
            id: <T = Account | null>(value: ResultRecord["id"]) => T;
            ronin: ReducedFunction & {
                createdAt: <T = Account | null>(value: ResultRecord["ronin"]["createdAt"]) => T;
                createdBy: <T = Account | null>(value: ResultRecord["ronin"]["createdBy"]) => T;
                updatedAt: <T = Account | null>(value: ResultRecord["ronin"]["updatedAt"]) => T;
                updatedBy: <T = Account | null>(value: ResultRecord["ronin"]["updatedBy"]) => T;
            };
            name: <T = Account | null>(value: string) => T;
            email: <T = Account | null>(value: string) => T;
        };
    };
    /* Get multiple account records */
    accounts: ReducedFunction & {
        <T = Accounts>(options?: Partial<CombinedInstructions>): T;
        after: <T = Accounts>(value: CombinedInstructions["after"]) => T;
        before: <T = Accounts>(value: CombinedInstructions["before"]) => T;
        including: <T = Accounts>(value: CombinedInstructions["including"]) => T;
        limitedTo: <T = Accounts>(value: CombinedInstructions["limitedTo"]) => T;
        orderedBy: <T = Accounts>(value: CombinedInstructions["orderedBy"]) => T;
        selecting: <T = Accounts>(value: CombinedInstructions["selecting"]) => T;
        using: <T = Accounts>(value: CombinedInstructions["using"]) => T;
        with: ReducedFunction & {
            <T = Accounts>(options: CombinedInstructions["with"]): T;
            id: <T = Accounts>(value: ResultRecord["id"]) => T;
            ronin: ReducedFunction & {
                createdAt: <T = Accounts>(value: ResultRecord["ronin"]["createdAt"]) => T;
                createdBy: <T = Accounts>(value: ResultRecord["ronin"]["createdBy"]) => T;
                updatedAt: <T = Accounts>(value: ResultRecord["ronin"]["updatedAt"]) => T;
                updatedBy: <T = Accounts>(value: ResultRecord["ronin"]["updatedBy"]) => T;
            };
            name: <T = Accounts>(value: string) => T;
            email: <T = Accounts>(value: string) => T;
        };
    };
};
"
`;

exports[`module query declaration constant with no modules 1`] = `
"const get: {};
"
`;

exports[`module query declaration constant with multiple models 1`] = `
"const get: {
    /* Get a single account record */
    account: ReducedFunction & {
        <T = Account | null>(options?: Partial<CombinedInstructions>): T;
        after: <T = Account | null>(value: CombinedInstructions["after"]) => T;
        before: <T = Account | null>(value: CombinedInstructions["before"]) => T;
        including: <T = Account | null>(value: CombinedInstructions["including"]) => T;
        limitedTo: <T = Account | null>(value: CombinedInstructions["limitedTo"]) => T;
        orderedBy: <T = Account | null>(value: CombinedInstructions["orderedBy"]) => T;
        selecting: <T = Account | null>(value: CombinedInstructions["selecting"]) => T;
        using: <T = Account | null>(value: CombinedInstructions["using"]) => T;
        with: ReducedFunction & {
            <T = Account | null>(options: CombinedInstructions["with"]): T;
            id: <T = Account | null>(value: ResultRecord["id"]) => T;
            ronin: ReducedFunction & {
                createdAt: <T = Account | null>(value: ResultRecord["ronin"]["createdAt"]) => T;
                createdBy: <T = Account | null>(value: ResultRecord["ronin"]["createdBy"]) => T;
                updatedAt: <T = Account | null>(value: ResultRecord["ronin"]["updatedAt"]) => T;
                updatedBy: <T = Account | null>(value: ResultRecord["ronin"]["updatedBy"]) => T;
            };
            name: <T = Account | null>(value: string) => T;
            email: <T = Account | null>(value: string) => T;
        };
    };
    /* Get multiple account records */
    accounts: ReducedFunction & {
        <T = Accounts>(options?: Partial<CombinedInstructions>): T;
        after: <T = Accounts>(value: CombinedInstructions["after"]) => T;
        before: <T = Accounts>(value: CombinedInstructions["before"]) => T;
        including: <T = Accounts>(value: CombinedInstructions["including"]) => T;
        limitedTo: <T = Accounts>(value: CombinedInstructions["limitedTo"]) => T;
        orderedBy: <T = Accounts>(value: CombinedInstructions["orderedBy"]) => T;
        selecting: <T = Accounts>(value: CombinedInstructions["selecting"]) => T;
        using: <T = Accounts>(value: CombinedInstructions["using"]) => T;
        with: ReducedFunction & {
            <T = Accounts>(options: CombinedInstructions["with"]): T;
            id: <T = Accounts>(value: ResultRecord["id"]) => T;
            ronin: ReducedFunction & {
                createdAt: <T = Accounts>(value: ResultRecord["ronin"]["createdAt"]) => T;
                createdBy: <T = Accounts>(value: ResultRecord["ronin"]["createdBy"]) => T;
                updatedAt: <T = Accounts>(value: ResultRecord["ronin"]["updatedAt"]) => T;
                updatedBy: <T = Accounts>(value: ResultRecord["ronin"]["updatedBy"]) => T;
            };
            name: <T = Accounts>(value: string) => T;
            email: <T = Accounts>(value: string) => T;
        };
    };
    /* Get a single post record */
    post: ReducedFunction & {
        <T = Post | null>(options?: Partial<CombinedInstructions>): T;
        after: <T = Post | null>(value: CombinedInstructions["after"]) => T;
        before: <T = Post | null>(value: CombinedInstructions["before"]) => T;
        including: <T = Post | null>(value: CombinedInstructions["including"]) => T;
        limitedTo: <T = Post | null>(value: CombinedInstructions["limitedTo"]) => T;
        orderedBy: <T = Post | null>(value: CombinedInstructions["orderedBy"]) => T;
        selecting: <T = Post | null>(value: CombinedInstructions["selecting"]) => T;
        using: <T = Post | null>(value: CombinedInstructions["using"]) => T;
        with: ReducedFunction & {
            <T = Post | null>(options: CombinedInstructions["with"]): T;
            id: <T = Post | null>(value: ResultRecord["id"]) => T;
            ronin: ReducedFunction & {
                createdAt: <T = Post | null>(value: ResultRecord["ronin"]["createdAt"]) => T;
                createdBy: <T = Post | null>(value: ResultRecord["ronin"]["createdBy"]) => T;
                updatedAt: <T = Post | null>(value: ResultRecord["ronin"]["updatedAt"]) => T;
                updatedBy: <T = Post | null>(value: ResultRecord["ronin"]["updatedBy"]) => T;
            };
            title: <T = Post | null>(value: string) => T;
            description: <T = Post | null>(value: string) => T;
        };
    };
    /* Get multiple post records */
    posts: ReducedFunction & {
        <T = Posts>(options?: Partial<CombinedInstructions>): T;
        after: <T = Posts>(value: CombinedInstructions["after"]) => T;
        before: <T = Posts>(value: CombinedInstructions["before"]) => T;
        including: <T = Posts>(value: CombinedInstructions["including"]) => T;
        limitedTo: <T = Posts>(value: CombinedInstructions["limitedTo"]) => T;
        orderedBy: <T = Posts>(value: CombinedInstructions["orderedBy"]) => T;
        selecting: <T = Posts>(value: CombinedInstructions["selecting"]) => T;
        using: <T = Posts>(value: CombinedInstructions["using"]) => T;
        with: ReducedFunction & {
            <T = Posts>(options: CombinedInstructions["with"]): T;
            id: <T = Posts>(value: ResultRecord["id"]) => T;
            ronin: ReducedFunction & {
                createdAt: <T = Posts>(value: ResultRecord["ronin"]["createdAt"]) => T;
                createdBy: <T = Posts>(value: ResultRecord["ronin"]["createdBy"]) => T;
                updatedAt: <T = Posts>(value: ResultRecord["ronin"]["updatedAt"]) => T;
                updatedBy: <T = Posts>(value: ResultRecord["ronin"]["updatedBy"]) => T;
            };
            title: <T = Posts>(value: string) => T;
            description: <T = Posts>(value: string) => T;
        };
    };
};
"
`;
